src/test/java/com/tabnine/binary/BootstrapperSupportTests.java

```diff
@@ -0,0 +1,80 @@
+        preferences.clear();
+    }
+
+    @Test
+    public void testWhenBootstrapperVersionIsNotLocalItWillDownloadIt() throws Exception {
+        when(binaryRemoteSource.fetchPreferredVersion(StaticConfig.getTabNineBundleVersionUrl())).thenReturn(Optional.of("9.9.9"));
+        when(bundleDownloader.downloadAndExtractBundle("9.9.9")).thenReturn(Optional.of(new BinaryVersion("9.9.9")));
+        assertThat(binaryVersionFetcher.fetchBinary(), equalTo(versionFullPath("9.9.9")));
+    }
+    @Test
+    public void testWhenBootstrapperVersionExistsItWillUseIt() throws Exception {
+        Preferences preferences = Preferences.userNodeForPackage(BootstrapperSupport.class);
+        preferences.put("bootstrapped version","8.8.8");
+        List<BinaryVersion> binaryVersions = aVersions();
+        binaryVersions.add(new BinaryVersion("8.8.8"));
+        when(localBinaryVersions.listExisting()).thenReturn(binaryVersions);
+        assertThat(binaryVersionFetcher.fetchBinary(), equalTo(versionFullPath("8.8.8")));
+    }
+    @Test
+    public void testWhenGreaterVersionThanPreferedBootstrapperVersionExistsItWillBeUsedInstead() throws Exception {
+        Preferences preferences = Preferences.userNodeForPackage(BootstrapperSupport.class);
+        preferences.put("bootstrapped version","5.5.5");
+        List<BinaryVersion> binaryVersions = aVersions();
+        binaryVersions.add(new BinaryVersion("55.55.55"));
+        when(localBinaryVersions.listExisting()).thenReturn(binaryVersions);
+        assertThat(binaryVersionFetcher.fetchBinary(), equalTo(versionFullPath("55.55.55")));
+    }
+    @Test
+    public void testWhenBootstrappedVersionDidNotExistLocallyTheBootstrapperWillDownloadAgain() throws Exception {
+        Preferences preferences = Preferences.userNodeForPackage(BootstrapperSupport.class);
+        preferences.put("bootstrapped version","6.6.6");
+        when(localBinaryVersions.listExisting()).thenReturn(aVersions());
+        when(binaryRemoteSource.fetchPreferredVersion(StaticConfig.getTabNineBundleVersionUrl())).thenReturn(Optional.of("66.66.66"));
+        when(bundleDownloader.downloadAndExtractBundle("66.66.66")).thenReturn(Optional.of(new BinaryVersion("66.66.66")));
+        assertThat(binaryVersionFetcher.fetchBinary(), equalTo(versionFullPath("66.66.66")));
+    }
+    @Test
+    public void testWhenBootstrappedVersionFailedToDownloadWillFallbackToLocalVersion() throws Exception {
+        when(localBinaryVersions.listExisting()).thenReturn(aVersions());
+        when(binaryRemoteSource.fetchPreferredVersion(StaticConfig.getTabNineBundleVersionUrl())).thenReturn(Optional.of("7.7.7"));
+        when(binaryRemoteSource.fetchPreferredVersion()).thenReturn(Optional.of(PREFERRED_VERSION));
+        when(bundleDownloader.downloadAndExtractBundle("7.7.7")).thenReturn(Optional.empty());
+        assertThat(binaryVersionFetcher.fetchBinary(), equalTo(versionFullPath(PREFERRED_VERSION)));
+    }
```

Comment:
I would add a test here to make sure that when multiple versions exist, it will pick the latest

Reply:
localBinaryVersions.listExisting() returns them ordered by latest